### ISSUE TITLE
insulation score computation in _insul_diamond_dense could be wrong

### ISSUE THREAD

**Issue by dmalzl** [open2c/cooltools#163]
I am currently trying to figure out why the insulation score computed from the cooltools `_find_insulation_boundaries_dense` looks so odd compared to published data. Sifting through the code I discovered that the above mentioned function uses another one `_insul_diamond_dense` which I assume tries to compute the insulation score according to [Crane et al. 2015](https://www.nature.com/articles/nature14450). However, the way the indices for the insulation diamond are computed is wrong. According to Crane et al. (and described here https://github.com/dekkerlab/crane-nature-2015/blob/master/scripts/matrix2insulation.pl) indices for the diamond are computed as:
```
lo = i - window
hi = i + window + 1
diamond = m[lo: i, i + 1: hi]
```
In contrast, `_insu_diamond_dense` computes these indices as
```
lo = i - window - 1
hi = i + window
diamond = m[lo: i + 1, i: hi]
```

which selects the wrong elements of the matrix as shown below.
```
window = 2

#matrix elements selected according to index computation in Crane et al. 2015
[[0, 0, 0, 1, 1, 0, 0, 0, 0, 0],
 [0, 0, 0, 1, 2, 2, 0, 0, 0, 0],
 [0, 0, 0, 0, 2, 3, 3, 0, 0, 0],
 [0, 0, 0, 0, 0, 3, 4, 4, 0, 0],
 [0, 0, 0, 0, 0, 0, 4, 5, 5, 0],
 [0, 0, 0, 0, 0, 0, 0, 5, 6, 6],
 [0, 0, 0, 0, 0, 0, 0, 0, 6, 6],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

#matrix elements selected according to index computation _insul_diamond_dense
[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 1, 1, 0, 0, 0, 0, 0, 0],
 [0, 0, 1, 2, 2, 0, 0, 0, 0, 0],
 [0, 0, 0, 2, 3, 3, 0, 0, 0, 0],
 [0, 0, 0, 0, 3, 4, 4, 0, 0, 0],
 [0, 0, 0, 0, 0, 4, 5, 5, 0, 0],
 [0, 0, 0, 0, 0, 0, 5, 6, 6, 0],
 [0, 0, 0, 0, 0, 0, 0, 6, 6, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]
```

**Comment by mimakaev** [comment]
I would not call it "wrong" because we never explicitly mention that we are trying to compute the score exactly the way it's done in Crane et al.  Also, diamond scores appeared before Crane et al:  I first remember them in  Sofueva et al, and I've been using them since I heard her talk in the Spring of 2013. 

It brings up an important point of what should be the bounds of the diamond, and which approach is more correct. There are two potential issues: including the main/first diagonal, and whether interactions of the current bin should be in the diamond. 

As I understand, our diamond includes one element of the main diagonal and two elements of the first diagonal, but by default we exclude the first two diagonals. Crane et al starts with the second diagonal. We're even here, but our insulation allows for ignoring arbitrary # of diagonals. 

Crane does not include upstream or downstream interactions of the current bin in the insulation score. We do include both. Which one is better and why? I don't have an opinion here. Maybe we should have it as an option?

**Comment by dmalzl** [comment]
You are right. I shouldn't have used such a harsh word and I just assumed that the score is computed like described in by Crane et al. because the procedures look so similar. Anyway, I am quite new to the field and just trying to get a grip on the methodology. I just wanted to point it out because I noticed the discrepancy between the methods.

**Comment by mimakaev** [comment]
It's a valid point though whether we should include interactions of the bin in the diamond. Do you have examples computed using two approaches? Does one of them seem more "correct" visually?

**Comment by dmalzl** [comment]
The plots below show a comparison of both approaches for some predefined TAD boundaries I called with HiCExplorer and were computed on a 20kb matrix with some 400M valid ditags. The code to generate the plots is based on the cooltools example (https://cooltools.readthedocs.io/en/latest/notebooks/05_insulation-score.html). For the crane example I simply modified the `_insul_diamond_dense` function as follows:
```
def _insul_diamond_dense_crane(mat, window=10, ignore_diags=2, norm_by_median=True):
    """
    Calculates the insulation score of a Hi-C interaction matrix.

    Parameters
    ----------
    mat : numpy.array
        A dense square matrix of Hi-C interaction frequencies.
        May contain nans, e.g. in rows/columns excluded from the analysis.

    window : int
        The width of the window to calculate the insulation score.

    ignore_diags : int
        If > 0, the interactions at separations < `ignore_diags` are ignored
        when calculating the insulation score. Typically, a few first diagonals
        of the Hi-C map should be ignored due to contamination with Hi-C
        artifacts.

    norm_by_median : bool
        If True, normalize the insulation score by its NaN-median.

    """
    if ignore_diags:
        mat = mat.copy()
        for i in range(-ignore_diags + 1, ignore_diags):
            numutils.set_diag(mat, np.nan, i)

    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        #cooltools code
        #N = mat.shape[0]
        #score = np.nan * np.ones(N)
        #for i in range(0, N):
        #    lo = max(0, i + 1 - window)
        #    hi = min(i + window, N)
            # nanmean of interactions to reduce the effect of bad bins
        #    score[i] = np.nanmean(mat[lo : i + 1, i:hi])
        
        #crane code
        N = mat.shape[0]
        score = np.nan * np.ones(N)
        for i in range(0 + window, N - window + 1):
            lo = i - window
            hi = i + window + 1
            # nanmean of interactions to reduce the effect of bad bins
            score[i] = np.nanmean(mat[lo : i, i + 1: hi])
        if norm_by_median:
            score /= np.nanmedian(score)
    return score
```
Mean profiles were computed using `np.nanmean`. 
In summary the Crane et al. approach seems to yield lower insulation scores but if this really makes a difference for comparing conditions or calling insulating loci I cannot say
![image](https://user-images.githubusercontent.com/52696584/82153544-5c40e900-9868-11ea-9a6d-b4af16475466.png)
![image](https://user-images.githubusercontent.com/52696584/82153548-67941480-9868-11ea-8668-a8945bb9fa34.png)

**Comment by dmalzl** [comment]
short edit:
a little difference is observable when comparing the insulation on compartment boarders (as defined by signflips of the first eigenvector. With cooltools there is an offset in comparing two conditions (both plots in a given image is a comparison between a WT and treatment) while with Crane this is not the case.

cooltools:
![image](https://user-images.githubusercontent.com/52696584/82154445-46362700-986e-11ea-9f2e-d48991f16186.png)

Crane:
![image](https://user-images.githubusercontent.com/52696584/82154453-5c43e780-986e-11ea-84c6-dde955edee44.png)

**Comment by golobor** [comment]
Hi, @dmalzl ,
many thanks for looking into this detailed comparison!!
I was not aware of the difference in the exact shape of diamonds between Crane and our implementation. A few years ago, I've been also playing with different diamonds, but could not find any noticeable difference in performance. However, tests like the one you have provided are very interesting!
(a) I agree that it would be nice to specify different parameters of the shape of the diamond. Unfortunately, I won't be able to fix the code in the nearest future. But, if you're interested, I could share some ideas re:implementation (shouldn't be too difficult)!
(b) the plots you have provided differ in y-scale roughly two-fold, which could exaggerate the difference between the two scores. With that said, even after the correction some differences would remain. But, then again, the biggest difference between the two conditions is just a shift by a fixed factor.

**Comment by dmalzl** [comment]
Hi Anton,

thanks for the reply. I actually tried to exclude more diagonals but the result remained the same more or less. However, I tried to normalize the score by the mean of scores instead of the median and this has resolved the offset, as the offset also remained with the Crane et al. version. 

You are lucky because I dug into the code more than I wanted due to some compatibility issues with coolers converted from HiCExplorers h5 format so I know quite a bit of it already :) If you would share your ideas I could have a look if you want me to.

Best regards,
Daniel

**Comment by golobor** [comment]
ok, upon re-reading of this, it seems that the difference between the Crane and our implementation is that, in the latter, the diamond is offset away from the diagonal by 1 pixel, while, in our version, the lower left corner of the diamond overlaps the diagonal.
@dmalzl , it also seems that I dropped this ball and never followed up on the idea of implementing a more flexible version of insulation scores! Feel free to get in touch if you're still interested ! closing this issue for now.
